src/main/scala/mesosphere/marathon/core/appinfo/impl/DefaultInfoService.scala

```diff
@@ -0,0 +1,79 @@
+package mesosphere.marathon.core.appinfo.impl
+
+import mesosphere.marathon.core.appinfo.AppInfo
+import mesosphere.marathon.core.appinfo._
+import mesosphere.marathon.state._
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+import scala.collection.immutable.Seq
+
+private[appinfo] class DefaultInfoService(
+    groupManager: GroupManager,
+    appRepository: AppRepository,
+    newBaseData: () => AppInfoBaseData) extends AppInfoService with GroupInfoService {
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def queryForAppId(id: PathId, embed: Set[AppInfo.Embed]): Future[Option[AppInfo]] = {
+    log.debug(s"queryForAppId $id")
+    appRepository.currentVersion(id).flatMap {
+      case Some(app) => newBaseData().appInfoFuture(app, embed).map(Some(_))
+      case None      => Future.successful(None)
+    }
+  }
+
+  override def queryAll(selector: AppSelector, embed: Set[AppInfo.Embed]): Future[Seq[AppInfo]] = {
+    log.debug(s"queryAll")
+    groupManager.rootGroup()
+      .map(_.transitiveApps.filter(selector.matches))
+      .flatMap(resolveAppInfos(_, embed))
+  }
+
+  override def queryAllInGroup(groupId: PathId, embed: Set[AppInfo.Embed]): Future[Seq[AppInfo]] = {
+    log.debug(s"queryAllInGroup $groupId")
+    groupManager
+      .group(groupId)
+      .map(_.map(_.transitiveApps).getOrElse(Seq.empty))
+      .flatMap(resolveAppInfos(_, embed))
+  }
+
+  override def queryForGroup(group: Group,
```

Comment:
The other methods are careful to prevent round-trips to the taskTracker and other query methods -- which also has the advantage of providing at a consistent snapshots of deployments, tasks. By using recursion here, you are requerying this information multiple times. I guess that this information will be eventually polled in the UI so that performance optimization should be worth it.